src/astarte-client/transforms/dataTree.ts

```diff
@@ -0,0 +1,355 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import _ from 'lodash';
+
+import { AstarteInterface } from '../models';
+import type {
+  AstarteData,
+  AstarteDataType,
+  AstarteDataValue,
+  AstarteDataTuple,
+  AstartePropertyData,
+  AstarteDatastreamIndividualData,
+  AstarteDatastreamObjectData,
+  AstarteInterfaceValues,
+  AstarteIndividualDatastreamInterfaceValue,
+  AstarteAggregatedDatastreamInterfaceValue,
+} from '../types';
+
+const getEndpointDataType = (iface: AstarteInterface, endpoint: string): AstarteDataType => {
+  const matchedMapping = AstarteInterface.findEndpointMapping(iface, endpoint);
+  if (matchedMapping == null) {
+    throw new Error(`Could not find an interface mapping for the endpoint ${endpoint}`);
+  }
+  return matchedMapping.type;
+};
+
+const isAstarteDataValue = (value: unknown): value is AstarteDataValue =>
+  !_.isUndefined(value) && (!_.isPlainObject(value) || _.isNull(value));
+
+const isPropertiesInterfaceValue = (value: unknown): value is AstarteDataValue =>
+  isAstarteDataValue(value);
+
+const isIndividualDatastreamInterfaceValue = (
+  value: unknown,
+): value is AstarteIndividualDatastreamInterfaceValue => isAstarteDataValue(_.get(value, 'value'));
+
+const isAggregatedDatastreamInterfaceValue = (
+  value: unknown,
+): value is AstarteAggregatedDatastreamInterfaceValue => Array.isArray(value);
+
+type AstarteDataTreeKind = 'properties' | 'datastream_object' | 'datastream_individual';
+
+const getDataTreeKind = (iface: AstarteInterface): AstarteDataTreeKind => {
+  if (iface.type === 'properties') {
+    return 'properties';
+  }
+  if (iface.aggregation === 'object') {
+    return 'datastream_object';
+  }
+  return 'datastream_individual';
+};
+
+interface AstarteDataTreeLeafParams<
+  Data extends AstartePropertyData | AstarteDatastreamIndividualData | AstarteDatastreamObjectData
+> {
+  interface: AstarteInterface;
+  data: Data extends AstarteDatastreamObjectData
+    ? AstarteDatastreamObjectData[]
+    : Data extends AstarteDatastreamIndividualData
+    ? AstarteDatastreamIndividualData[]
+    : AstartePropertyData;
+  endpoint?: string;
+  relativeEndpoint?: string;
+  parentNode?: AstarteDataTreeBranch<Data> | null;
+}
+
+class AstarteDataTreeLeaf<
+  Data extends AstartePropertyData | AstarteDatastreamIndividualData | AstarteDatastreamObjectData
+> {
+  readonly parent: AstarteDataTreeBranch<Data> | null;
+
+  readonly endpoint: string;
+
+  readonly relativeEndpoint: string;
+
+  readonly dataKind: AstarteDataTreeKind;
+
+  readonly data: Data extends AstarteDatastreamObjectData
+    ? AstarteDatastreamObjectData[]
+    : Data extends AstarteDatastreamIndividualData
+    ? AstarteDatastreamIndividualData[]
+    : AstartePropertyData;
+
+  readonly linearizedData: AstarteData[];
```

Comment:
This has no reference to what kind of data is stored in the tree. We do have this information (reflected using typescript generics). Why removing this information to the user?
`AstarteData` has an optional `timestamp` which we know is not used for properties

Reply:
As in a reply to a comment above, I've clearly distinguished the two cases (timestamp / no timestamp) and reflected the two cases when linearizing trees using Typescript generics.